README.md
# shis

## what's this
shis is a command-line tool.  
Fast and quickly search of bash history.

The rough function is as follows.  
```
$ history | grep [keyword]
```

## How to use this
This cmd application was developed by go-lang.  
So you need to get or generate a binary to run.

release page is [here](https://github.com/shogysd/shis/releases)

### How to make binary
Please execute this command on macOS or Linux with go-lang installed.  
( If you are using another operating system (e.g. Windows)? Sorry, this application is not supported. )  

#### get source files
Download the source files from [https://github.com/shogysd/shis/archive/master.zip](https://github.com/shogysd/shis/archive/master.zip).  
( If you want to use git command, please type this command. )
```
// download by git
$ git clone https://github.com/shogysd/shis.git
```

#### build source files
Go to the project root directory and execute the command.
```
$ make build
```
If the build is successful, the binary is generated here -> `/bin/[Arch]/shis` .

### How to run it
plwase type this!
```
$ shis [search target]
// e.g. $ shis ls
```
When you type this command, get the help message.
```
$ shis -h
```

## ATTENTION
To use all features, please make the following settings.

- `PROMPT_COMMAND='history -a'`
  - This command line tool gets history from file.  
  Therefore, to perform an accurate history search, you must write the history to a file each time you run the command.
- `HISTTIMEFORMAT='[Arbitrary set value]'`
  - Time stamps must be saved for accurate time to appear in search results.


CHANGELOG
Version 0.0
    - Migrate from old repository


test/argparser_test.go
package test

import (
	"../shis"
	"testing"
)

func TestArgParser(t *testing.T) {

	var err error
	var errCode int
	var words []string
	var args = []string{}
	var histFileName = "testHome"
	var invertFlag = false
	var dupAllowFlag = false
	var maxRec = 100
	var andScanFlag = true
	var color = true
	/*
		var histories []string
		var sigTreeHead = shis.NewDataCV("")
		var histLines []shis.HistData
		var latestSigNo = -1
	*/

	args = []string{"arg1", "arg2", "arg3"}
	if err, errCode, words = shis.ArgParser(args, &histFileName, &invertFlag, &dupAllowFlag, &maxRec, &andScanFlag, &color); err != nil {
		t.Error("target list make err")
	} else {

		if errCode != 0 {
			t.Error("target list make err")
		}

		for i, word := range words {
			if word != words[i] {
				t.Error("target list make err")
				break
			}
		}

	}

}


test/histoyscanner_test.go
package test

import (
	"../shis"
	"testing"
)

func TestHistoryScanner(t *testing.T) {

	var err error
	var errCode int
	var words []string
	var invertFlag = false
	var dupAllowFlag = false
	var maxRec = 100
	var andScanFlag = true
	var histories []string
	var sigTreeHead = shis.NewDataCV("")
	var histLines []shis.HistData
	var latestSigNo = -1

	words = []string{"01", "02"}
	latestSigNo = len(words)
	histories = []string{"history line 01", "history line 02", "history line 03"}

	if err, errCode = shis.MakeTree(&words, sigTreeHead); err != nil || errCode != 0 {
		t.Error("sig tree err")
	}

	if err, errCode = shis.HistoryScanner(&histories, sigTreeHead, invertFlag, dupAllowFlag, &histLines, &maxRec, andScanFlag, latestSigNo); err != nil || errCode != 0 {
		t.Error("scan history scan err")
	}

}


test/histfilereader_test.go
package test

import (
	"../shis"
	"testing"
)

func TestHistFileReader(t *testing.T) {

	var err error
	var errCode int
	var histories = []string{}
	var histFileName = ""

	histFileName = "./testsamples/sample_history_file"
	err, errCode = shis.HistFileReader(&histories, &histFileName)
	if err != nil || errCode != 0 {
		t.Error("file open fail")
	}

	histFileName = "./testsamples/sample_history_file_ZZZ"
	err, errCode = shis.HistFileReader(&histories, &histFileName)
	if err == nil || errCode == 0 {
		t.Error("file open fail ()")
	}

	histories = []string{}
	histFileName = "./testsamples/sample_history_file"
	err, errCode = shis.HistFileReader(&histories, &histFileName)
	if err != nil || errCode != 0 {
		t.Error("file open fail")
	}
	var hisAnswer = []string{"#0", "command line 01", "#1", "command line 02", "#3", "command line 03"}
	for i, readLine := range histories {
		if readLine != hisAnswer[i] {
			t.Error("read file test err")
			break
		}
	}
}


shis/argparser.go
package shis

import (
	"errors"
	"strconv"
)

func ArgParser(args []string, histFileName *string, invertFlag *bool, dupAllowFlag *bool, maxRec *int, andScanFlag *bool, color *bool) (error, int, []string) {

	var words []string
	var andScanReaded = false

	if len(args) == 1 {
		return errors.New("asked for help"), -1, nil
	} else {
		i := 1
		for i < len(args) {

			if string(args[i][0]) == "-" {

				switch args[i] {

				case "-v":
					*invertFlag = true

				case "-d":
					*dupAllowFlag = true

				case "-n":
					if i < len(args)-1 {
						var err error
						if *maxRec, err = strconv.Atoi(args[i+1]); err != nil {
							return errors.New("arguments parse err"), 101, nil
						}
						i += 1
					} else {
						return errors.New("arguments parse err"), 101, nil
					}

				case "-c":
					if i < len(args)-1 {
						if args[i+1] == "true" || args[i+1] == "t" {
							*color = true
						} else if args[i+1] == "false" || args[i+1] == "f" {
							*color = false
						} else {
							return errors.New("arguments parse err"), 101, nil
						}
						i += 1
					} else {
						return errors.New("arguments parse err"), 101, nil
					}

				case "-f":
					if i < len(args)-1 {
						*histFileName = args[i+1]
						i += 1
					} else {
						return errors.New("arguments parse err"), 101, nil
					}

				case "-a":
					if andScanReaded == false {
						andScanReaded = true
						*andScanFlag = true
					} else {
						return errors.New("arguments parse err"), 101, nil
					}

				case "-o":
					if andScanReaded == false {
						andScanReaded = true
						*andScanFlag = false
					} else {
						return errors.New("arguments parse err"), 101, nil
					}

				case "-h":
					return errors.New("asked for help"), -1, nil

				case "--version":
					return errors.New("require version print"), -2, nil

				default:
					return errors.New(string(args[i]) + " arguments parse err"), 101, nil
				}

			} else {
				words = append(words, args[i])
			}
			i += 1

		}
		return nil, 0, words
	}
}


shis/scanengine.go
package shis

type DataCV struct {
	// data is single character string, it's data
	Data string
	// termination flag
	Term int
	// next node (this flag is used only when end flag ia false)
	NextNode *DataCV
	// parallel node (this flag is used only when end flag ia false)
	ParallelNode *DataCV
	// jump node
	JumpNode *DataCV
}

func NewDataCV(data string) *DataCV {
	cont := new(DataCV)
	cont.Data = data
	cont.Term = -1
	cont.NextNode = nil
	cont.ParallelNode = nil
	cont.JumpNode = nil
	return cont
}

func addSig(sigs *[]string, sigTreeHead *DataCV) (error, int) {

	for sigNo, sig := range *sigs {
		pNode := sigTreeHead
		eNode := NewDataCV("")
		for i, char := range sig {
			nNode := NewDataCV("")
			if pNode.NextNode == nil {
				nNode = NewDataCV(string(char))
				pNode.NextNode = nNode
				pNode = pNode.NextNode
			} else {
				eNode = pNode.NextNode
				for {
					if eNode.Data == string(char) {
						pNode = eNode
						break
					} else if eNode.ParallelNode != nil {
						eNode = eNode.ParallelNode
					} else {
						nNode = NewDataCV(string(char))
						eNode.ParallelNode = nNode
						pNode = eNode.ParallelNode
						break
					}
				}
			}
			if i == len(sig)-1 {
				pNode.Term = sigNo
			}
		}
	}
	return nil, 0
}

func addJumpRoad(sigs *[]string, sigTreeHead *DataCV) (error, int) {
	// make tree (add jump load)
	for _, sig := range *sigs {
		pNode := sigTreeHead
		eNode := NewDataCV("")
		jNode := NewDataCV("")

		for _, char := range sig {
			eNode = pNode.NextNode

			if pNode == sigTreeHead {
				for {
					if eNode.Data == string(char) {
						eNode.JumpNode = sigTreeHead
						pNode = eNode
						break
					} else if eNode.ParallelNode != nil {
						eNode = eNode.ParallelNode
					}
				}
			} else {
				jNode = pNode.JumpNode.NextNode
				for {
					if eNode.Data == string(char) {
						if jNode.Data == eNode.Data { // == string(char)
							eNode.JumpNode = jNode
							pNode = eNode
							break
						} else if jNode.ParallelNode != nil {
							jNode = jNode.ParallelNode
						} else {
							eNode.JumpNode = sigTreeHead
							pNode = eNode
							break
						}
					} else if eNode.ParallelNode != nil {
						eNode = eNode.ParallelNode
					} else {
						eNode.JumpNode = sigTreeHead
						pNode = eNode.NextNode
						break
					}
				}
			}
		}
	}
	return nil, 0
}

func moreSig(sig *string, sigTreeHead *DataCV, sigNo int) (error, int) {

	pNode := sigTreeHead
	eNode := NewDataCV("")

	for i, char := range *sig {
		nNode := NewDataCV("")
		if pNode.NextNode == nil {
			nNode = NewDataCV(string(char))
			pNode.NextNode = nNode
			pNode = pNode.NextNode
		} else {
			eNode = pNode.NextNode
			for {
				if eNode.Data == string(char) {
					pNode = eNode
					break
				} else if eNode.ParallelNode != nil {
					eNode = eNode.ParallelNode
				} else {
					nNode = NewDataCV(string(char))
					eNode.ParallelNode = nNode
					pNode = eNode.ParallelNode
					break
				}
			}
		}
		if i == len(*sig)-1 {
			pNode.Term = sigNo
		}
	}

	return nil, 0
}

func moreJumpRoad(sig *string, sigTreeHead *DataCV) (error, int) {

	pNode := sigTreeHead
	eNode := NewDataCV("")
	jNode := NewDataCV("")

	for _, char := range *sig {
		eNode = pNode.NextNode

		if pNode == sigTreeHead {
			for {
				if eNode.Data == string(char) {
					eNode.JumpNode = sigTreeHead
					pNode = eNode
					break
				} else if eNode.ParallelNode != nil {
					eNode = eNode.ParallelNode
				}
			}
		} else {
			jNode = pNode.JumpNode.NextNode
			for {
				if eNode.Data == string(char) {
					if jNode.Data == eNode.Data { // == string(char)
						eNode.JumpNode = jNode
						pNode = eNode
						break
					} else if jNode.ParallelNode != nil {
						jNode = jNode.ParallelNode
					} else {
						eNode.JumpNode = sigTreeHead
						pNode = eNode
						break
					}
				} else if eNode.ParallelNode != nil {
					eNode = eNode.ParallelNode
				} else {
					eNode.JumpNode = sigTreeHead
					pNode = eNode.NextNode
					break
				}
			}
		}
	}

	return nil, 0
}

func MakeTree(words *[]string, sigTreeHead *DataCV) (error, int) {

	addSig(words, sigTreeHead)
	addJumpRoad(words, sigTreeHead)

	return nil, 0
}

func AddTree(word *string, sigTreeHead *DataCV, latestSigNo int) (error, int) {

	latestSigNo += 1
	moreSig(word, sigTreeHead, latestSigNo)
	moreJumpRoad(word, sigTreeHead)

	return nil, 0
}

func flagsChecker(flags []bool) bool {
	for _, f := range flags {
		if f == false {
			return false
		}
	}
	return true
}

func SearchTree(word *string, sigTreeHead *DataCV, andScanFlag bool, latestSigNo int) (error, int, bool) {
	/*
	* ret arg
	* ( error, int(err code), bool(hit or not hit) )
	 */

	var tapFlags = make([]bool, latestSigNo)

	if sigTreeHead.NextNode == nil {
		// tree is not constructed
		return nil, 0, false
	}

	pNode := sigTreeHead
	eNode := NewDataCV("")

	for _, char := range *word {
		eNode = pNode.NextNode

		for {
			if eNode.Data == string(char) {
				if eNode.Term != -1 {
					tapFlags[eNode.Term] = true
					if !andScanFlag || (andScanFlag && flagsChecker(tapFlags)) {
						return nil, 0, true
					} else {
						break
					}
				} else if eNode.JumpNode != nil && eNode.JumpNode.Term != -1 {
					tapFlags[eNode.JumpNode.Term] = true
					if !andScanFlag || (andScanFlag && flagsChecker(tapFlags)) {
						return nil, 0, true
					} else {
						break
					}
				} else {
					pNode = eNode
					break
				}
			} else if eNode.ParallelNode != nil {
				eNode = eNode.ParallelNode
			} else if pNode.JumpNode != nil {
				pNode = pNode.JumpNode
				eNode = pNode.NextNode
			} else {
				pNode = sigTreeHead
				break
			}
		}
	}
	// searched till the end but did not hit
	return nil, 0, false
}


shis/historyscanner.go
package shis

import (
	"os"
	"strconv"
	"strings"
)

type HistData struct {
	Ts           int64
	HistLineData string
}

func NewHist() *HistData {
	cont := new(HistData)
	cont.Ts = -999999
	cont.HistLineData = ""
	return cont
}

func HistoryScanner(
	histories *[]string, sigTreeHead *DataCV, invertFlag bool, dupAllowFlag bool, histLines *[]HistData, maxRec *int, andScanFlag bool, latestSigNo int) (error, int) {

	// history scan
	i := len(*histories) - 1
	recCounter := 0
	insertHist := NewHist()
	for i >= 0 {
		if (*histories)[i] != "" && string((*histories)[i][0]) != "#" {
			// hitFlag := false

			searchLine := strings.TrimRight((*histories)[i], " ")

			if err, errCode, hitFlag := SearchTree(&(searchLine), sigTreeHead, andScanFlag, latestSigNo); err != nil {
				switch errCode {
				case 401:
					// return errors.New("arguments parse err"), 401
				}
			} else if hitFlag == !invertFlag {

				if i >= 2 && string((*histories)[i-1][0]) == "#" {
					if ts, err := strconv.ParseInt((*histories)[i-1][1:], 10, 64); err != nil {
						// timestamo parse err
						os.Exit(1)
					} else {
						insertHist.Ts = ts
						insertHist.HistLineData = searchLine
						// insertHist.HistLineData = (*histories)[i]
					}
				} else {
					insertHist.Ts = -999999
					insertHist.HistLineData = searchLine
					// insertHist.HistLineData = (*histories)[i]
				}

				hitLineDupFlag := false // Match search criteria

				if dupAllowFlag == false {

					for _, hitLine := range *histLines {
						if hitLine.HistLineData == insertHist.HistLineData {
							hitLineDupFlag = true
							break
						}
					}

					if hitLineDupFlag == false {
						*histLines = append(*histLines, *insertHist)
					}

				} else {
					*histLines = append(*histLines, *insertHist)
				}
				recCounter += 1
				if recCounter >= *maxRec {
					break
				}
			}
		}
		i -= 1
	}
	return nil, 0
}


cmd/main.go
package main

import (
	"../shis"
	"fmt"
	"os"
	"strings"
	"time"
)

var versionString string = "shis: Version information not found"

func main() {

	errMessFormat := "err: %s [code:%d]\n"

	var err error
	var errCode int
	var words []string
	var histFileName = os.Getenv("HOME") + "/" + ".bash_history"
	var invertFlag = false
	var dupAllowFlag = false
	var maxRec = 100
	var andScanFlag = true
	var color = true
	var histories []string
	var sigTreeHead = shis.NewDataCV("")
	var histLines []shis.HistData
	var latestSigNo = -1

	// parse of arguments
	if err, errCode, words = shis.ArgParser(os.Args, &histFileName, &invertFlag, &dupAllowFlag, &maxRec, &andScanFlag, &color); err != nil {

		switch errCode {
		case -1:
			// error code '-1' mean asked for help by user
			shis.HelpPrinter()
			os.Exit(0)
		case -2:
			// error code '-2' mean asked for version
			fmt.Println(versionString)
			os.Exit(0)
		case 101:
			// error code '101' mean arguments parse err
			// commandline latest argument is '-v'
			fmt.Printf(errMessFormat, err, errCode)
			fmt.Printf("\nPlease see below for how to use the command.\n'shis -h'\n")
			os.Exit(1)
		}
	}

	latestSigNo = len(words)

	// make sig tree
	if err, errCode = shis.MakeTree(&words, sigTreeHead); err != nil {
		switch errCode {
		case 201:
			fmt.Printf(errMessFormat, err, errCode)
			os.Exit(1)
		}
	}

	// histry file read
	if err, errCode = shis.HistFileReader(&histories, &histFileName); err != nil {
		switch errCode {
		case 301:
			// error code '301' mean histFile not found
			// if use -f option you must to use file path for history file
			fmt.Printf(errMessFormat, err, errCode)
			os.Exit(1)
		}
	}

	// host scanner
	if err, errCode = shis.HistoryScanner(&histories, sigTreeHead, invertFlag, dupAllowFlag, &histLines, &maxRec, andScanFlag, latestSigNo); err != nil {
		switch errCode {
		case 301:
			// error code '301' mean histFile not found
			// if use -f option you must to use file path for history file
			fmt.Printf(errMessFormat, err, errCode)
			os.Exit(1)
		}
	}

	// hist printer
	i := len(histLines) - 1
	for i >= 0 {

		if histLines[i].Ts == -999999 {
			fmt.Printf(" No date information    ")
		} else {
			fmt.Printf(" %s   ", time.Unix(histLines[i].Ts, 0).Format("01-02-2006 (15:04:05)"))
			// fmt.Printf(" %s   ", time.Unix(histLines[i].Ts, 0).Format("Jan-_2-2006 (15:04:05)"))
		}

		if color == true {
			outStr := histLines[i].HistLineData
			for _, word := range words {
				outStr = strings.Replace(outStr, word, "\x1b[1;31m"+word+"\x1b[0;39m", -1)
			}
			fmt.Printf("%s\n", outStr)
		} else {
			fmt.Printf("%s\n", histLines[i].HistLineData)
		}

		i -= 1

	}

	os.Exit(0)
}
